sklearn/kernel_approximation.py

```diff
@@ -22,7 +24,169 @@
+
+        self.degree = degree
+        self.coef0 = coef0
+        self.n_components = n_components
+        self.random_state = random_state
+
+    def fit(self, X, y=None):
+        """Fit the model with X.
+
+        Initializes the internal variables. The method needs no information
+        about the distribution of data, so we only care about n_features in X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data, where n_samples in the number of samples
+            and n_features is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns the transformer.
+        """
+
+        X = check_array(X, accept_sparse="csc")
+        random_state = check_random_state(self.random_state)
+
+        n_features = X.shape[1]
+        if self.coef0 != 0:
+            n_features += 1
+
+        self.indexHash_ = random_state.randint(0, high=self.n_components,
+                                               size=(self.degree, n_features))
+
+        self.bitHash_ = random_state.choice(a=[-1, 1],
+                                            size=(self.degree, n_features))
+        return self
+
+    def transform(self, X):
+        """Generate the feature map approximation for X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            New data, where n_samples in the number of samples
+            and n_features is the number of features.
+
+        Returns
+        -------
+        X_new : array-like, shape (n_samples, n_components)
+        """
+
+        check_is_fitted(self)
+        X = check_array(X, accept_sparse="csc")
+
+        if sp.issparse(X) and self.coef0 != 0:
+            X = sp.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))],
+                          format="csc")
+
+        elif not sp.issparse(X) and self.coef0 != 0:
+            X = np.hstack([X, np.sqrt(self.coef0)*np.ones((X.shape[0], 1))])
+
+        if X.shape[1] != self.indexHash_.shape[1]:
+            raise ValueError("Number of features of test samples does not"
+                             " match that of training samples.")
+
+        count_sketches = np.zeros((X.shape[0], self.degree, self.n_components))
+
+        if sp.issparse(X):
+            for j in range(X.shape[1]):
+                for d in range(self.degree):
+                    iHashIndex = self.indexHash_[d, j]
+                    iHashBit = self.bitHash_[d, j]
+                    count_sketches[:, d, iHashIndex] += \
+                        (iHashBit * X[:, j]).toarray().ravel()
+
+        else:
+            for j in range(X.shape[1]):
+                for d in range(self.degree):
+                    iHashIndex = self.indexHash_[d, j]
+                    iHashBit = self.bitHash_[d, j]
+                    count_sketches[:, d, iHashIndex] += iHashBit * X[:, j]
```

Comment:
I tried to remove one loop (`j` or `d`) and do it vectorized. Vectorizing `d` like
```python
count_sketches = np.zeros((X.shape[0], 3, n_components))
for j in range(X.shape[1]):
    iHashIndex = self.indexHash_[:, j]
    iHashBit = self.bitHash_[:, j]
    count_sketches[:, np.arange(degree), iHashIndex] += iHashBit * X[:, j, np.newaxis]
```
was slower. Vectorizing `j` is tricky because `iHashIndex` can point to the same index for different values of `j`.
So let's keep it as is. Possible improvements maybe in later PRs.